UPGRADE-4.0.md

```diff
@@ -0,0 +1,207 @@
+
+### Removed
+
+- `ecs-instance-profile` sub-module has been removed; this functionality is available through the [`terraform-aws-modules/terraform-aws-autoscaling`](https://github.com/terraform-aws-modules/terraform-aws-autoscaling) module starting with version [v6.5.0](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/pull/194). Please see the [`examples/ec2`](https://github.com/terraform-aws-modules/terraform-aws-ecs/tree/master/examples/ec2) example for a demonstration on how to use and integrate with the `terraform-aws-autoscaling` module.
+
+### Variable and output changes
+
+1. Removed variables:
+
+    - `default_capacity_provider_strategy` is now incorporated into the `fargate_capacity_providers` and `autoscaling_capacity_providers` variables.
+
+2. Renamed variables:
+
+    - `create_ecs` -> `create`
+    - `name` -> `cluster_name`
+
+3. Added variables:
+
+    - `cluster_configuration` has been added under a dynamic block with all current attributes supported
+
+4. Removed outputs:
+
+    - `ecs_cluster_name`
+
+5. Renamed outputs:
+
+    - `ecs_cluster_id` -> `cluster_id`
+    - `ecs_cluster_arn` -> `cluster_arn`
+
+6. Added outputs:
+
+    - `cluster_capacity_providers`
+    - `autoscaling_capacity_providers`
+
+## Upgrade Migrations
+
+### Before v3.x Example
+
+```hcl
+module "ecs" {
+  source  = "terraform-aws-modules/ecs/aws"
+  version = "3.5.0"
+
+  name               = "example"
+  container_insights = true
+
+  capacity_providers = ["FARGATE", "FARGATE_SPOT", aws_ecs_capacity_provider.prov1.name]
+
+  default_capacity_provider_strategy = [{
+    capacity_provider = aws_ecs_capacity_provider.prov1.name
+    weight            = "1"
+  }]
+}
+
+module "ec2_profile" {
+  source  = "terraform-aws-modules/ecs/aws//modules/ecs-instance-profile"
+  version = "3.5.0"
+
+  name = local.name
+}
+
+resource "aws_ecs_capacity_provider" "prov1" {
+  name = "prov1"
+
+  auto_scaling_group_provider {
+    auto_scaling_group_arn = module.autoscaling.autoscaling_group_arn
+  }
+}
+```
+
+### After v4.x Example
+
+```hcl
+module "ecs" {
+  source  = "terraform-aws-modules/ecs/aws"
+  version = "4.0.0"
+
+  cluster_name = "example"
+
+  fargate_capacity_providers = {
+    "FARGATE"      = {}
+    "FARGATE_SPOT" = {}
+  }
+
+  autoscaling_capacity_providers = {
+    prov1 = {
+      auto_scaling_group_arn = module.autoscaling.autoscaling_group_arn
+      default_capacity_provider_strategy = {
+        weight = 1
+      }
+    }
+  }
+}
+
+module "ec2_profile" {
+  source  = "terraform-aws-modules/ecs/aws//modules/ecs-instance-profile"
+  # Users can pin and stay on v3.5.0 until they able to use the IAM instance
+  # profile provided through the autoscaling group module
+  version = "3.5.0"
```

Comment:
I did try to migrate the instance role and profile from the sub-module in v3.x to the autoscaling group, but the instance profile path and name is different from the role and currently the autoscaling group module makes the assumption that the IAM role and instance profile names and paths are the same. Users can pin their profile to v3.5.0 and continue to use without issue and avoid any disruption when upgrading